x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/tree/NodeSubclassTests.java

```diff
@@ -618,45 +622,77 @@ protected boolean hasAtLeastTwoChildren(Class<? extends Node<?>> toBuildClass) {
         for (String path: paths) {
             Path root = PathUtils.get(path);
             int rootLength = root.toString().length() + 1;
-            Files.walkFileTree(root, new SimpleFileVisitor<Path>() {
-
-                @Override
-                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
-                    if (Files.isRegularFile(file) && file.getFileName().toString().endsWith(".class")) {
-                        String className = file.toString();
-                        // Chop off the root and file extension
-                        className = className.substring(rootLength, className.length() - ".class".length());
-                        // Go from "path" style to class style
-                        className = className.replace(PathUtils.getDefaultFileSystem().getSeparator(), ".");
-
-                        // filter the class that are not interested
-                        // (and IDE folders like eclipse)
-                        if (!className.startsWith("org.elasticsearch.xpack.ql") && !className.startsWith("org.elasticsearch.xpack.sql")) {
-                            return FileVisitResult.CONTINUE;
-                        }
-
-                        Class<?> c;
-                        try {
-                            c = Class.forName(className);
-                        } catch (ClassNotFoundException e) {
-                            throw new IOException("Couldn't find " + file, e);
-                        }
 
-                        if (false == Modifier.isAbstract(c.getModifiers())
-                                && false == c.isAnonymousClass()
-                                && clazz.isAssignableFrom(c)) {
-                            Class<? extends T> s = c.asSubclass(clazz);
-                            results.add(s);
+            // load classes from jar files
+            // NIO FileSystem API is not used since it trips the SecurityManager
+            // https://bugs.openjdk.java.net/browse/JDK-8160798
+            // so iterate the jar "by hand"
+            if (path.endsWith(".jar") && path.contains("x-pack-ql")) {
+                try (JarInputStream jar = jarStream(root)) {
+                    JarEntry je = null;
+                    while ((je = jar.getNextJarEntry()) != null) {
+                        String name = je.getName();
+                        if (name.endsWith(".class")) {
+                            String className = name.substring(0, name.length() - ".class".length()).replace("/", ".");
+                            maybeLoadClass(clazz, className, root + "!/" + name, results);
```

Comment:
Is `!/` intended as a visual marker in the path? The `location` is later only used for the message in the Exception, so I guess so, but curious about it.

Reply:
No, it's a special marker inside the URL to indicate the path inside the archive:
`jar:/my/file.jar!/org/my/package/clazz.class`